### PR TITLE
Allow record valued counters

### PR DESCRIPTION
**Pull request by rgrinberg**
This allows us to record more than one time series at once.  Use this for the GC
stats.

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

### CODE REVIEW

**Comment by rgrinberg** [comment]
@emillon as  a side note, why is this library named "Catapult"?

**Comment by aalekseyev** [comment]
I expanded the comment section at the top so at least the relationship to the catapult project is clear. I don't know why that project shares a name with a siege weapon, though.

File: src/catapult/catapult.mli [hunk 1]
```diff
@@ -25,7 +43,7 @@ val on_process_start : t -> program:string -> args:string list -> event
 val on_process_end : t -> event -> unit
 
 (** Emit a counter event. This is measuring the value of an integer variable. *)
```
Comment (aalekseyev) [inline src/catapult/catapult.mli:45]: Why do we call this `emit_counter` if it's not even a number in general?
Should it be called `emit_event`, or something like that?

**Review by aalekseyev — APPROVED** [review]

File: src/catapult/catapult.mli [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
+
+    Trace viewer is a part of the catapult project
+    (https://github.com/catapult-project/catapult/blob/master/tracing/README.md).
+
+    The trace format is documented at:
+    https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview
+*)
+
+module Json : sig
+  (** Poor man's json type *)
```
Comment (emillon) [inline src/catapult/catapult.mli:12]:
```suggestion
  (** Simplified JSON type *)
```

File: src/catapult/catapult.mli (hunk 1 above)
Comment (emillon) [reply src/catapult/catapult.mli:45]: This is terminology from the trace format. Counter events (type "C") can have several series in them. There are other kinds of events (not counters) if we wanted to annotate the trace with phases for example ("evaluating rules", "executing rules", etc)

File: test/blackbox-tests/test-cases/trace-file.t/run.t [hunk 3]
```diff
@@ -16,17 +16,7 @@ This captures the commands that are being run:
...
-  {"name": "major_words", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ....00}}
-  {"name": "minor_collections", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ...}}
-  {"name": "minor_words", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ....00}}
-  {"name": "promoted_words", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ....00}}
-  {"name": "stack_size", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ...}}
-  {"name": "top_heap_words", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value": ...}}
+  $ <trace.json grep '"C"' | cut -c 2- | sed -E 's/([^0-9])[0-9]+/\1.../g' | sort -u
+  {"name": "evaluated-rules", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value":...}}
+  {"name": "fds", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"value":...}}
+  {"name": "gc", "pid": ..., "tid": ..., "ph": "C", "ts": ..., "args": {"live_words":...,"free_words":...,"stack_size":...,"heap_words":...,"top_heap_words":...,"minor_words":....,"major_words":....,"promoted_words":....,"compactions":...,"major_collections":...,"minor_collections":...}}
```
Comment (emillon) [inline test/blackbox-tests/test-cases/trace-file.t/run.t:22]: It's nice that we can but different series in a counter. I don't think that was supported when I first added that.

**Comment by emillon** [comment]
> @emillon as a side note, why is this library named "Catapult"?

I struggled to find the name of that file format. I found the docs on https://chromium.googlesource.com/catapult which:

> Catapult is the home for several performance tools that span from gathering, displaying and analyzing performance data.

If we were to pick a format today I think [CTF](https://diamon.org/ctf/) would be a better choice - multicore uses that and there are actual OCaml libraries to emit it.

**Comment by rgrinberg** [comment]
> If we were to pick a format today I think CTF would be a better choice - multicore uses that and there are actual OCaml libraries to emit it.

Okay, what do you think about renaming the library to `Ctf` then?

File: src/catapult/catapult.ml [hunk 4]
```diff
@@ -80,32 +149,31 @@ let on_process_end t (id, name) =
...
-
-let emit_counter_float =
-  let float () f = Printf.sprintf "%.2f" f in
-  fun t key value -> gen_emit_counter t key float value
+  let values =
+    Buffer.clear t.buffer;
+    Json.to_buf (Json.Object values) t.buffer;
+    Buffer.contents t.buffer
+  in
+  printf t {|{"name": %S, "pid": 0, "tid": 0, "ph": "C", "ts": %s, "args": %s}|}
```
Comment (rgrinberg) [inline src/catapult/catapult.ml:159]: Did you check if an arbitrary json value is allowed for `"args"`? The documentation wasn't very clear on this.

**Comment by aalekseyev** [comment]
@rgrinberg, CTF is a different format.

**Comment by rgrinberg** [comment]
>  CTF is a different format.

I immediately assumed that stood for "Chrome Trace Format". Thanks for the correction 😆 

Okay, how about `Tef`. The document reverse to this as a "Trace Event Format".

**Comment by aalekseyev** [comment]
The catapult own docs [refer](https://github.com/catapult-project/catapult/blob/master/tracing/README.md) to it as "Chrome's trace_event format", so it sounds like there's no great name.

Making up our own abbreviations seems confusing to me. We could call it Chrome_trace_event_format.
I think catapult is a fine name, though. It's not like anybody is confused about which part of catapult project we're talking about.

**Comment by rgrinberg** [comment]
I wouldn't mind `Chrome_trace`, but I agree that `Catapult` can remain

**Comment by snowleopard** [comment]
I would prefer `Chrome_trace` actually since it's immediately obvious, whereas I always need to google `Catapult` when I see it.

**Comment by emillon** [comment]
> I immediately assumed that stood for "Chrome Trace Format".

To complicate things further... https://github.com/KDAB/ctf2ctf

`Chrome_trace` sounds good to me too.

**Pull request merged**